design-documents/frontend/server-side-rendering.md

```diff
@@ -0,0 +1,129 @@
+# Server-Side Rendering (SSR)
+
+_This document refers to SSR in the context of a React web application (e.g., PWA). The concept of SSR does not apply to native apps._
+
+_SSR belongs to the BFF layer of [service isolation](https://github.com/magento/architecture/tree/master/design-documents/service-isolation)._
+
+SSR can be a challenge to implement correctly, but the benefits have the potential to provide the unparalleled performance you seek to decrease bounce rate and drive more business.
+
+## Pros
+
+- Faster, first meaningful paint.
+- No need to render initial content via JavaScript on the client.
+- SEO/bot support.
+- Better SEO performance.
+- JavaScript code can be shared universally (full-stack).
+
+## Cons
+
+- User may see some content before it's interactive.
+- Difficult or tedious to implement.
+- May require additional configuration and refactoring.
+- Some features [not yet supported](#not-supported-yet).
+
+## How it works
+
+The basic idea is that we don't want the web browser (client) to be responsible for rendering content via JavaScript, because JavaScript is notoriously slow, especially at rendering DOM and especially on older or mobile devices. Unnecessary rendering also consumes more energy, which has the potential to suck a mobile device's battery dry. :battery:
+
+If we leverage the HTTP server to do as much heavy lifting as we can, some of this heavy lifting (non-sensitive pages) can be cached as well, resulting in a faster, better user experience.
+
+### Without SSR
+
+1. A web browser (client) requests a URL for a particular route (e.g., `/`).
+1. The HTTP server runs the appropriate code for that route.
+   - HTML is partially generated with empty containers that need to be filled and rendered by JavaScript after the initial page load (bad for SEO).
+1. The web browser (client) receives and processes the HTML, but the user may see some empty content that hasn't been filled in yet.
+1. JavaScript is loaded and processed, including the React library.
+1. Ideally, a single GraphQL API request is made to collect the data required to fill in the empty containers with content.
```

Comment:
Think this is a bit too specific:

> Ideally, a single GraphQL API request is made to collect the data required to fill in the empty containers with content

Can we make this description bit more general? A single API request for an entire UI is not ideal because it doesn't differentiate between high and low priority content, so critical functionality gets delayed due to network time for low priority content.

Maybe just say:

> The client-side JavaScript makes a request to one or several APIs for the initial data needed to render the page